modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java

```diff
@@ -0,0 +1,224 @@
+            array.setTop(rawAfterKey.values);
+        }
+        this.keys = new TreeMap<>(array::compare);
+        this.canEarlyTerminate = Arrays.stream(sources)
+            .allMatch(CompositeValuesSourceConfig::canEarlyTerminate);
+    }
+
+    boolean canEarlyTerminate() {
+        return canEarlyTerminate;
+    }
+
+    private int[] getReverseMuls() {
+        return Arrays.stream(sources).mapToInt(CompositeValuesSourceConfig::reverseMul).toArray();
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long zeroBucket) throws IOException {
+        assert zeroBucket == 0L;
+
+        // Replay all documents that contain at least one top bucket (collected during the first pass).
+        grow(keys.size()+1);
+        for (LeafContext context : contexts) {
+            DocIdSetIterator docIdSetIterator = context.docIdSet.iterator();
+            if (docIdSetIterator == null) {
+                continue;
+            }
+            final CompositeValuesSource.Collector collector =
+                array.getLeafCollector(context.ctx, getSecondPassCollector(context.subCollector));
+            int docID;
+            while ((docID = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                collector.collect(docID);
+            }
+        }
+
+        int num = Math.min(size, keys.size());
+        final InternalComposite.InternalBucket[] buckets = new InternalComposite.InternalBucket[num];
+        final int[] reverseMuls = getReverseMuls();
+        int pos = 0;
+        for (int slot : keys.keySet()) {
+            CompositeKey key = array.toCompositeKey(slot);
+            InternalAggregations aggs = bucketAggregations(slot);
+            int docCount = bucketDocCount(slot);
+            buckets[pos++] = new InternalComposite.InternalBucket(key, reverseMuls, docCount, aggs);
+        }
+        return new InternalComposite(name, size, Arrays.asList(buckets), reverseMuls, pipelineAggregators(), metaData());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        final int[] reverseMuls = getReverseMuls();
+        return new InternalComposite(name, size, Collections.emptyList(), reverseMuls, pipelineAggregators(), metaData());
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (leaf != null) {
+            leaf.docIdSet = builder.build();
+            contexts.add(leaf);
+        }
+        leaf = new LeafContext(ctx, sub);
+        builder = new RoaringDocIdSet.Builder(ctx.reader().maxDoc());
+        final CompositeValuesSource.Collector inner = array.getLeafCollector(ctx, getFirstPassCollector());
+        return new LeafBucketCollector() {
+            @Override
+            public void collect(int doc, long zeroBucket) throws IOException {
+                assert zeroBucket == 0L;
+                inner.collect(doc);
+            }
+        };
+    }
```

Comment:
I think its worth adding some JAvaDocs here to explain how the collection works and what is happening in the first pass and the second pass

Reply:
++